### ISSUE TITLE
ONNX export of torch.jit.script module fails

### ISSUE THREAD

**Issue by julianmack** [pytorch/pytorch#33495]
## 🐛 Bug

I'm trying to export an ONNX model with the following elements:
1) Control flow 
2) An LSTM

My understanding is that I need to `torch.jit.script` the model before passing it to `torch.onnx.export` in order to accurately represent the control flow as described here: https://pytorch.org/docs/stable/onnx.html#tracing-vs-scripting.

However, the ONNX export operation throws a RuntimeError (same as #24824). 

I am aware of the issue #24235 which has overlap- but the apparent conclusion in this thread is to avoid scripting with LSTMs which prevents the use of 1) and 2) at the same time. 

## To Reproduce

Minimal example here:
```python
import torch
from typing import Tuple
from torch import Tensor
from copy import deepcopy

HIDDEN_SIZE = 4
NUM_LAYERS = 1

class SimpleControlFlow(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.lstm = torch.nn.LSTM(HIDDEN_SIZE, HIDDEN_SIZE, NUM_LAYERS)

    def forward(self, x: Tensor, hn: Tensor, cn: Tensor, 
                sos: Tensor,
    ) -> Tuple[Tensor, Tensor, Tensor]:
        out, (hn, cn) = self.lstm(x, (hn, cn))
        if sos:
            out += 1.
        return out, hn, cn

def gen_args(batch, seq_len, sos, hidden_size=HIDDEN_SIZE, num_layers=NUM_LAYERS):
    """Generates input args for SimpleControlFlow."""
    y = torch.randn(seq_len, batch, hidden_size)
    zeros = torch.zeros(num_layers, batch, hidden_size) # hn == cn == zeros
    return y, zeros, zeros, torch.tensor([sos])

model = SimpleControlFlow()
```

### Aside: demonstrate scripting is required
```python
args1 = gen_args(batch=1, seq_len=3, sos=False)
args2 = gen_args(batch=1, seq_len=1, sos=True)
model_t = torch.jit.trace(deepcopy(model), args1, check_trace=True)
out1_nn = model(*args1)
out1_trace = model_t(*args1)
out2_nn = model(*args2)
out2_trace = model_t(*args2)

assert torch.allclose(out1_nn[0], out1_trace[0]), "Will pass as args1 used for tracing.\n"
assert not torch.allclose(out2_nn[0], out2_trace[0]), "Will pass as `sos` was traced as constant = False.\n"
```
As expected, this throws a `TracerWarning` and the assertions pass.
### Aside end
```python
args = gen_args(batch=1, seq_len=3, sos=False)

model = torch.jit.script(model)

example_outputs = model(*args) # i.e. this runs 


input_names = ['y', 'hn_in', 'cn_in', 'sos']
output_names = ['out', 'hn_out', 'cn_out']
dynamic_axes = {
    "y": {0: "seq_len", 1: "batch"},
    "hn_in": {1: "batch"},
    "cn_in": {1: "batch"},
    "out": {0: "seq_len", 1: "batch"},
    "hn_out": {1: "batch"},
    "cn_out": {1: "batch"},
}

torch.onnx.export(
    model,
    args,
    'model.onnx',
    export_params=True,
    verbose=True,
    example_outputs=example_outputs,
    dynamic_axes=None,
    input_names=input_names,
    output_names=output_names,
    opset_version=11,
    )
```
Raises 
```shell
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-a255a7548070> in <module>
     28     input_names=input_names,
     29     output_names=output_names,
---> 30     opset_version=11,
     31     )

~/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/onnx/__init__.py in export(model, args, f, export_params, verbose, training, input_names, output_names, aten, export_raw_ir, operator_export_type, opset_version, _retain_param_name, do_constant_folding, example_outputs, strip_doc_string, dynamic_axes, keep_initializers_as_inputs)
    146                         operator_export_type, opset_version, _retain_param_name,
    147                         do_constant_folding, example_outputs,
--> 148                         strip_doc_string, dynamic_axes, keep_initializers_as_inputs)
    149 
    150 

~/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/onnx/utils.py in export(model, args, f, export_params, verbose, training, input_names, output_names, aten, export_raw_ir, operator_export_type, opset_version, _retain_param_name, do_constant_folding, example_outputs, strip_doc_string, dynamic_axes, keep_initializers_as_inputs)
     64             _retain_param_name=_retain_param_name, do_constant_folding=do_constant_folding,
     65             example_outputs=example_outputs, strip_doc_string=strip_doc_string,
---> 66             dynamic_axes=dynamic_axes, keep_initializers_as_inputs=keep_initializers_as_inputs)
     67 
     68 

~/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/onnx/utils.py in _export(model, args, f, export_params, verbose, training, input_names, output_names, operator_export_type, export_type, example_outputs, propagate, opset_version, _retain_param_name, do_constant_folding, strip_doc_string, dynamic_axes, keep_initializers_as_inputs, fixed_batch_size)
    414                                                         example_outputs, propagate,
    415                                                         _retain_param_name, do_constant_folding,
--> 416                                                         fixed_batch_size=fixed_batch_size)
    417 
    418         # TODO: Don't allocate a in-memory string for the protobuf

~/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/onnx/utils.py in _model_to_graph(model, args, verbose, training, input_names, output_names, operator_export_type, example_outputs, propagate, _retain_param_name, do_constant_folding, _disable_torch_constant_prop, fixed_batch_size)
    263         assert example_outputs is not None, "example_outputs must be provided when exporting a ScriptModule"
    264         try:
--> 265             method_graph, params = torch._C._jit_pass_lower_graph(model.forward.graph, model._c)
    266             in_vars, in_desc = torch.jit._flatten(tuple(args) + tuple(params))
    267             graph = _propagate_and_assign_input_shapes(

RuntimeError: isTensor() INTERNAL ASSERT FAILED at /opt/conda/conda-bld/pytorch_1579040055865/work/aten/src/ATen/core/ivalue_inl.h:90, please report a bug to PyTorch. Expected Tensor but got Bool (toTensor at /opt/conda/conda-bld/pytorch_1579040055865/work/aten/src/ATen/core/ivalue_inl.h:90)
frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x47 (0x7f81bec8a627 in /home/julian/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/lib/libc10.so)
frame #1: torch::jit::LowerGraph(torch::jit::Graph&, c10::intrusive_ptr<c10::ivalue::Object, c10::detail::intrusive_target_default_null_type<c10::ivalue::Object> > const&) + 0x674 (0x7f81c4670794 in /home/julian/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/lib/libtorch.so)
frame #2: <unknown function> + 0x6b6479 (0x7f81eff2b479 in /home/julian/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #3: <unknown function> + 0x28ba06 (0x7f81efb00a06 in /home/julian/miniconda3/envs/ml_ci/lib/python3.7/site-packages/torch/lib/libtorch_python.so)
frame #4: _PyMethodDef_RawFastCallKeywords + 0x264 (0x55f54c4e26e4 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #5: _PyCFunction_FastCallKeywords + 0x21 (0x55f54c4e2801 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #6: _PyEval_EvalFrameDefault + 0x4e8c (0x55f54c53e2bc in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #7: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #8: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #9: _PyEval_EvalFrameDefault + 0x14ce (0x55f54c53a8fe in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #10: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #11: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #12: _PyEval_EvalFrameDefault + 0x14ce (0x55f54c53a8fe in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #13: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #14: _PyFunction_FastCallKeywords + 0x325 (0x55f54c4e19c5 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #15: _PyEval_EvalFrameDefault + 0x4aa9 (0x55f54c53ded9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #16: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #17: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #18: _PyEval_EvalFrameDefault + 0x14ce (0x55f54c53a8fe in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #19: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #20: PyEval_EvalCodeEx + 0x44 (0x55f54c4803c4 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #21: PyEval_EvalCode + 0x1c (0x55f54c4803ec in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #22: <unknown function> + 0x1e004d (0x55f54c54904d in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #23: _PyMethodDef_RawFastCallKeywords + 0xe9 (0x55f54c4e2569 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #24: _PyCFunction_FastCallKeywords + 0x21 (0x55f54c4e2801 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #25: _PyEval_EvalFrameDefault + 0x4755 (0x55f54c53db85 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #26: _PyGen_Send + 0x2a2 (0x55f54c4db672 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #27: _PyEval_EvalFrameDefault + 0x1a6d (0x55f54c53ae9d in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #28: _PyGen_Send + 0x2a2 (0x55f54c4db672 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #29: _PyEval_EvalFrameDefault + 0x1a6d (0x55f54c53ae9d in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #30: _PyGen_Send + 0x2a2 (0x55f54c4db672 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #31: _PyMethodDef_RawFastCallKeywords + 0x8c (0x55f54c4e250c in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #32: _PyMethodDescr_FastCallKeywords + 0x4f (0x55f54c4e286f in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #33: _PyEval_EvalFrameDefault + 0x4c4c (0x55f54c53e07c in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #34: _PyFunction_FastCallKeywords + 0xfb (0x55f54c4e179b in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #35: _PyEval_EvalFrameDefault + 0x416 (0x55f54c539846 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #36: _PyFunction_FastCallKeywords + 0xfb (0x55f54c4e179b in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #37: _PyEval_EvalFrameDefault + 0x6a0 (0x55f54c539ad0 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #38: _PyEval_EvalCodeWithName + 0x2f9 (0x55f54c47f4f9 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #39: _PyFunction_FastCallDict + 0x400 (0x55f54c480800 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #40: _PyObject_Call_Prepend + 0x63 (0x55f54c497c43 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #41: PyObject_Call + 0x6e (0x55f54c48c95e in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #42: _PyEval_EvalFrameDefault + 0x1e20 (0x55f54c53b250 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #43: _PyEval_EvalCodeWithName + 0x5da (0x55f54c47f7da in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #44: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #45: _PyEval_EvalFrameDefault + 0x14ce (0x55f54c53a8fe in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #46: <unknown function> + 0x171cc6 (0x55f54c4dacc6 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #47: <unknown function> + 0x171ecb (0x55f54c4daecb in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #48: _PyMethodDef_RawFastCallKeywords + 0xe9 (0x55f54c4e2569 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #49: _PyCFunction_FastCallKeywords + 0x21 (0x55f54c4e2801 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #50: _PyEval_EvalFrameDefault + 0x4755 (0x55f54c53db85 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #51: _PyEval_EvalCodeWithName + 0x5da (0x55f54c47f7da in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #52: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #53: _PyEval_EvalFrameDefault + 0x6a0 (0x55f54c539ad0 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #54: <unknown function> + 0x171cc6 (0x55f54c4dacc6 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #55: <unknown function> + 0x171ecb (0x55f54c4daecb in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #56: _PyMethodDef_RawFastCallKeywords + 0xe9 (0x55f54c4e2569 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #57: _PyCFunction_FastCallKeywords + 0x21 (0x55f54c4e2801 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #58: _PyEval_EvalFrameDefault + 0x4755 (0x55f54c53db85 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #59: _PyEval_EvalCodeWithName + 0x5da (0x55f54c47f7da in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #60: _PyFunction_FastCallKeywords + 0x387 (0x55f54c4e1a27 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #61: _PyEval_EvalFrameDefault + 0x416 (0x55f54c539846 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #62: <unknown function> + 0x171cc6 (0x55f54c4dacc6 in /home/julian/miniconda3/envs/ml_ci/bin/python)
frame #63: <unknown function> + 0x171ecb (0x55f54c4daecb in /home/julian/miniconda3/envs/ml_ci/bin/python)
```

Note that this is raised despite the fact that all forward arguments, including the bool flag `sos` are `Tensors`.

## Expected behaviour
Expect ONNX model to export. 

## Environment
```shell
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1050 Ti with Max-Q Design
Nvidia driver version: 430.64
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.13.3
[pip3] torch-stft==0.1.4
[conda] _pytorch_select           0.1                       cpu_0    anaconda
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243    anaconda
[conda] mkl-service               2.3.0            py37he904b0f_0    anaconda
[conda] mkl_fft                   1.0.12           py37ha843d7b_0    anaconda
[conda] mkl_random                1.0.2            py37hd81dba3_0    anaconda
[conda] pytorch                   1.4.0           py3.7_cuda10.0.130_cudnn7.6.3_0    pytorch
[conda] torchaudio                0.4.0                    pypi_0    pypi
[conda] torchvision               0.2.1                    py37_0
```
## Additional context

This problem results from the `lstm` as opposed to the if statement - this can be seen by commenting out the if statement and addition in the `SimpleControlFlow` forward method. The control flow is present in the example as it necessary in my use-case.

cc @suo @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof

**Comment by BowenBao** [comment]
Hi @julianmack , thanks for reporting this. We are working on a fix to resolve this and #24235, #24824.

**Comment by julianmack** [comment]
Hi @suo @BowenBao @eellison: do you know which release you are expecting this to be fixed in?

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/42914

**Comment by BowenBao** [comment]
Should be fixed with #54112

**Comment by garymm** [comment]
Seems this works in PyTorch 1.10.0.
